### PR TITLE
Release Dashboard & Planning Analytics #159

### PR DESCRIPTION
**Pull request by codecharlan**
Objective: Implement API endpoints and calculation logic to support release dashboards and comprehensive planning analytics.

Data Dependencies from prerequisite: This issue relies on feature assignment data implemented in the prerequisite:
- Feature assignments via GET /api/releases/{releaseCode}/features
- planning_status field values: NOT_STARTED, IN_PROGRESS, BLOCKED, DONE
- feature_owner field for workload distribution
- blockage_reason field for blocked time analysis
- planned_completion_date field for timeline calculations
- notes field for audit trail

Requirements:

Dashboard API:

1. GET /api/releases/{code}/dashboard
   Release dashboard with overview, feature breakdown, progress metrics, health indicators, timeline data.

   Example Response:
   ```json
   {
     "releaseCode": "REL-2024-Q1",
     "releaseName": "Q1 2024 Release",
     "status": "IN_PROGRESS",
     "overview": {
       "totalFeatures": 45,
       "completedFeatures": 12,
       "inProgressFeatures": 18,
       "blockedFeatures": 3,
       "pendingFeatures": 12,
       "completionPercentage": 26.7,
       "estimatedDaysRemaining": 28
     },
     "healthIndicators": {
       "timelineAdherence": "DELAYED",
       "riskLevel": "MEDIUM",
       "blockedFeatures": 3
     },
     "timeline": {
       "startDate": "2024-01-15T00:00:00Z",
       "plannedEndDate": "2024-03-31T00:00:00Z",
       "estimatedEndDate": "2024-04-15T00:00:00Z",
       "actualEndDate": null
     },
     "featureBreakdown": {
       "byStatus": {
         "NOT_STARTED": 12,
         "IN_PROGRESS": 18,
         "BLOCKED": 3,
         "DONE": 12
       },
       "byOwner": {
         "user1@company.com": 5,
         "user2@company.com": 8,
         "unassigned": 3
       },
       "byPriority": {
         "CRITICAL": 3,
         "HIGH": 8,
         "MEDIUM": 6,
         "LOW": 4
       }
     }
   }
   ```

2. GET /api/releases/{code}/metrics
   Detailed metrics for analytics (completion rate, velocity, blocked time, workload distribution).

   Example Response:
   ```json
   {
     "releaseCode": "REL-2024-Q1",
     "releaseStatus": "IN_PROGRESS",
     "completionRate": 26.7,
     "velocity": {
       "featuresPerWeek": 3.2,
       "averageCycleTime": 8.5
     },
     "blockedTime": {
       "totalBlockedDays": 15,
       "averageBlockedDuration": 5.0,
       "percentageOfTime": 8.5,
       "blockageReasons": {
         "Technical Debt": 5,
         "Dependencies": 3
       }
     },
     "workloadDistribution": {
       "byOwner": [
         {
           "owner": "john.doe",
           "assignedFeatures": 8,
           "completedFeatures": 3,
           "inProgressFeatures": 4,
           "blockedFeatures": 1,
           "utilizationRate": 85.0
         }
       ]
     }
   }
   ```

Analytics API:

1. GET /api/planning/health
   Planning health report (releases by status, overdue/at-risk counts, planning accuracy).

   Example Response:
   ```json
   {
     "releasesByStatus": {
       "DRAFT": 8,
       "IN_PROGRESS": 12,
       "RELEASED": 45,
       "CANCELLED": 2
     },
     "atRiskReleases": {
       "overdue": 3,
       "criticallyDelayed": 2,
       "total": 5
     },
     "planningAccuracy": {
       "onTimeDelivery": 73.2,
       "averageDelay": 12.5,
       "estimationAccuracy": 78.9
     }
   }
   ```

2. GET /api/planning/trends
   Trend data (releases completed over time, average duration, planning accuracy).

   Example Response:
   ```json
   {
     "releasesCompleted": {
       "trend": [
         {"period": "2023-02", "value": 4},
         {"period": "2023-03", "value": 6},
         {"period": "2024-01", "value": 5}
       ],
       "total": 58
     },
     "averageReleaseDuration": {
       "trend": [
         {"period": "2023-02", "value": 89.5},
         {"period": "2023-03", "value": 92.3},
         {"period": "2024-01", "value": 87.1}
       ],
       "current": 87.1
     },
     "planningAccuracyTrend": {
       "onTimeDelivery": [
         {"period": "2023-02", "value": 68.2},
         {"period": "2023-03", "value": 71.5},
         {"period": "2024-01", "value": 73.2}
       ]
     }
   }
   ```

3. GET /api/planning/capacity
   Capacity planning data (workload by owner, commitments, overallocation warnings).

   Example Response:
   ```json
   {
     "overallCapacity": {
       "totalResources": 45,
       "utilizationRate": 87.3,
       "availableCapacity": 12.7,
       "overallocatedResources": 3
     },
     "workloadByOwner": [
       {
         "owner": "john.doe",
         "currentWorkload": 8,
         "capacity": 10,
         "utilizationRate": 80.0,
         "futureCommitments": 12,
         "overallocationRisk": "NONE"
       },
       {
         "owner": "jane.smith",
         "currentWorkload": 12,
         "capacity": 10,
         "utilizationRate": 120.0,
         "futureCommitments": 15,
         "overallocationRisk": "HIGH"
       }
     ],
     "commitments": {
       "activeReleases": 12,
       "plannedReleases": 8,
       "totalFeatures": 234,
       "estimatedEffort": 1580.5
     },
     "overallocationWarnings": [
       {
         "owner": "jane.smith",
         "severity": "HIGH",
         "overallocationPercentage": 120.0
       }
     ]
   }
   ```

Service Layer:
- Implement DashboardService for dashboard data aggregation and calculations
- Implement PlanningAnalyticsService for analytics calculations and insights

Test Coverage:
- Integration tests for all dashboard API endpoints
- Integration tests for all analytics API endpoints
- Tests for dashboard metric accuracy (completion rate, velocity, blocked time)
- Tests for timeline data inclusion in dashboard responses
- Tests for business logic validation (timeline adherence, risk level calculations)
- Tests for edge cases (releases with no features, all unassigned features)
- Tests for data consistency (feature counts add up, percentages are correct)
- Tests for realistic value ranges (velocity metrics within reasonable bounds)
- Tests for error handling (invalid release codes, missing data scenarios)

Acceptance Criteria:
- Release dashboard provides comprehensive metrics and visualizations
- Dashboard includes all required data: timeline, health indicators, feature breakdown, workload distribution
- Dashboard endpoints return accurate metrics (completion rate, velocity, blocked time)
- Planning analytics API delivers actionable insights
- Health reports accurately show releases by status and at-risk counts
- Trend data provides meaningful historical analysis
- Capacity planning identifies overallocation warnings correctly
- Planning accuracy is correctly calculated and reported
- All business logic rules are correctly implemented (timeline adherence, risk levels)
- Edge cases are handled gracefully (empty releases, invalid data)
- Field calculations are mathematically correct (percentages, rates, sums)
- Integration tests validate API behavior without implementation dependencies
- GET endpoints are public; write operations (POST, PUT, DELETE) require OAuth2 authentication

Implementation Details:

Business Logic Rules:

1. Timeline Adherence:
   - DELAYED if current date is greater than plannedEndDate and status is not RELEASED.
   - AT_RISK if within 14 days of plannedEndDate and completionPercentage is less than 70%.
   - ON_TRACK otherwise.

2. Risk Level:
   - HIGH if blockedFeatures is greater than 50% of totalFeatures or completionPercentage is less than 30%.
   - MEDIUM if blockedFeatures is greater than 25% of totalFeatures or completionPercentage is less than 60%.
   - LOW otherwise.

3. Velocity Calculation:
   - featuresPerWeek equals completed features divided by business weeks elapsed since release start.
     Business weeks = business days elapsed / 5 (Mon-Fri only, skip weekends). Do NOT use calendar days / 7.
   - estimatedDaysRemaining = ceil(remaining / featuresPerWeek) * 5.
     The multiplier is 5 (business days per week), not 7.
   - averageCycleTime equals the average days between feature creation and completion.
   - Requires at least 2 weeks of data; otherwise show 0.

4. At-Risk Releases:
   - overdue and criticallyDelayed are mutually exclusive; total = overdue + criticallyDelayed.
   - DRAFT releases use age-based classification (days since createdAt):
     - age > ft.planning.draft-critically-delayed-days → criticallyDelayed
     - age > ft.planning.draft-overdue-days (but <= critically-delayed threshold) → overdue
   - IN_PROGRESS (and other non-DRAFT, non-RELEASED, non-CANCELLED) releases use
     deadline-based classification: if now() is after plannedEndDate → overdue.
     IN_PROGRESS releases are never counted as criticallyDelayed.
   - RELEASED and CANCELLED releases are never counted as at-risk.

5. Blocked Time Calculation:
   - totalBlockedDays equals the sum of days each BLOCKED feature has been in that status.
   - percentageOfTime equals totalBlockedDays divided by total feature-days multiplied by 100.
   - blockageReasons is a map of blockage_reason values to counts, or an empty map if unavailable.

6. Workload Distribution:
   - utilizationRate equals completedFeatures divided by assignedFeatures multiplied by 100 for each owner.
   - Unassigned features (where feature_owner is null) should be counted under "unassigned" in the breakdown.
   - byPriority should contain counts of features by priority levels (CRITICAL, HIGH, MEDIUM, LOW).

7. Overallocation Risk Thresholds:
   - Use >= (inclusive) when comparing utilizationRate against thresholds.
   - utilizationRate >= ft.planning.overallocation-thresholds.high-threshold → HIGH
   - utilizationRate >= ft.planning.overallocation-thresholds.medium-threshold → MEDIUM
   - Otherwise → NONE

8. Release Dates:
   - plannedEndDate is calculated as 90 business days after release.createdAt.
   - estimatedEndDate for active releases is based on current velocity; for released releases it equals releasedAt.

Default Values and Edge Cases:
- Releases with no features should show 0 for all counts and percentages.
- Releases with no assigned features should show an "unassigned" count in breakdown.
- If no completed features exist, averageCycleTime should be 0.
- All percentages should be rounded to 1 decimal place.
- All calculations should handle division by zero gracefully.
- Released releases should always have timelineAdherence set to ON_TRACK and estimatedDaysRemaining set to 0.
- Features created without an explicit priority should default to MEDIUM.

FAIL_TO_PASS: com.sivalabs.ft.features.api.controllers.DashboardControllerIntegrationTests,com.sivalabs.ft.features.api.controllers.PlanningControllerIntegrationTests,com.sivalabs.ft.features.api.controllers.FeatureControllerTests#shouldCreateNewFeatureWithDefaultPriority

### CODE REVIEW

File: src/test/java/com/sivalabs/ft/features/api/controllers/FeatureControllerTests.java [hunk 1]
```diff
@@ -22,7 +22,7 @@ void shouldGetFeaturesByReleaseCode() {
                 .bodyJson()
                 .extractingPath("$.size()")
                 .asNumber()
-                .isEqualTo(2);
+                .isEqualTo(5);
```
[the remark this answers is not in the document]
Comment (codecharlan) [reply src/test/java/com/sivalabs/ft/features/api/controllers/FeatureControllerTests.java]: moved the required data into its class with clear comments of scenario

File: src/test/java/com/sivalabs/ft/features/api/controllers/FeatureControllerTests.java [hunk 2]
```diff
@@ -116,11 +116,11 @@ void shouldUpdateFeature() {
     @Test
     @WithMockOAuth2User(username = "user")
     void shouldDeleteFeature() {
-        var result = mvc.delete().uri("/api/features/{code}", "IDEA-2").exchange();
+        var result = mvc.delete().uri("/api/features/{code}", "IDEA-99").exchange();
         assertThat(result).hasStatusOk();
 
         // Verify deletion
-        var getResult = mvc.get().uri("/api/features/{code}", "IDEA-2").exchange();
+        var getResult = mvc.get().uri("/api/features/{code}", "IDEA-99").exchange();
```
[the remark this answers is not in the document]
Comment (codecharlan) [reply src/test/java/com/sivalabs/ft/features/api/controllers/FeatureControllerTests.java]: moved the required data into its class with clear comments of scenario

**Comment by DmitriyChernyakSetronica** [comment]
When running tests on the gold branch, I encountered a problem with the test:
```
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DashboardControllerIntegrationTests.shouldCalculateFeaturesPerWeekExactlyForKnownFixture:521 
expected: 5.0
 but was: 4.5
```

**Comment by codecharlan** [comment]
> When running tests on the gold branch, I encountered a problem with the test:
> 
> ```
> [INFO] 
> [INFO] Results:
> [INFO] 
> [ERROR] Failures: 
> [ERROR]   DashboardControllerIntegrationTests.shouldCalculateFeaturesPerWeekExactlyForKnownFixture:521 
> expected: 5.0
>  but was: 4.5
> ```

Screenshot 2026-02-25 at 03 52 49

I just ran the test again, and its isnt same experience, likely an inconsistent test data setup. I will look at it and fix it.

**Comment by codecharlan** [comment]
> When running tests on the gold branch, I encountered a problem with the test:
> 
> ```
> [INFO] 
> [INFO] Results:
> [INFO] 
> [ERROR] Failures: 
> [ERROR]   DashboardControllerIntegrationTests.shouldCalculateFeaturesPerWeekExactlyForKnownFixture:521 
> expected: 5.0
>  but was: 4.5
> ```

From my investigation, I was able to reproduce the issue by using another time zone; I have switched from using ZoneId.systemDefault() to ZoneOffset.UTC. The issue is resolved.

**Comment by DmitriyChernyakSetronica** [comment]
In my auto branch, my agent couldn't handle the calculation in this test; they couldn't restore the formula based on the conditions and visible test data. The requirements may need to be clarified—you'll need to check this.
``` 
[ERROR] Failures: 
[ERROR]   DashboardControllerIntegrationTests.shouldCalculateEstimatedDaysRemainingForActiveRelease:752 
expected: 50
 but was: 69
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by DmitriyChernyakSetronica** [comment]
PlanningControllerIntegrationTests#shouldApplyCorrectOverallocationRiskThresholds,PlanningControllerIntegrationTests#shouldIdentifyHighOverallocationForJaneSmith
To pass these tests, the agent used the requirements:
```
{
"owner": "jane.smith",
"currentWorkload": 12,
"capacity": 10,
"utilizationRate": 120.0,
"futureCommitments": 15,
"overallocationRisk": "HIGH"
}

And in overallocationWarnings:

{
"owner": "jane.smith",
"severity": "HIGH",
"overallocationPercentage": 120.0
}

This means that 120.0% should yield "HIGH". Therefore, >= should be used instead of >.
```
Is this the correct logic for "severity": "HIGH" - the threshold is >= 120%?

**Comment by DmitriyChernyakSetronica** [comment]
In my auto branch, my agent couldn't handle the calculation in this test; they couldn't restore the formula based on the conditions and visible test data. The requirements may need to be clarified—you'll need to check this.
PlanningControllerIntegrationTests.shouldCountAtLeastOneReleaseAsOverdue:256 
Did it correct the logic correctly?
Below is the agent's reasoning for reference:
```
The requirements were:

At-Risk Releases:

overdue - releases that are overdue (passed the planned end date)
criticallyDelayed - releases with a critical delay
total - the total number of at-risk releases
The requirements don't specify a specific threshold for "critically delayed." I used 30 days as a reasonable value, but tests show that this isn't appropriate for the test data.

Let's look at the underlying logic in the tests. Perhaps the categories should be interpreted differently:

overdue - all overdue releases (including critically delayed)
criticallyDelayed - a subset of overdue releases with a particularly large delay
But the test expects total = overdue + criticallyDelayed, which means disjoint categories.

Let's try a different approach - perhaps the threshold should be increased to match the test data. The comments indicate the releases were created in 2023, but it's now 2026, meaning a delay of years. Let's try 365 days (1 year).
```

**Comment by DmitriyChernyakSetronica** [comment]
My auto branch requires authentication to be configured in these tests.

**@WithMockOAuth2User**
class PlanningControllerIntegrationTests
Agent Reasoning
```This shows that:

All endpoints in the project require authentication (OAuth2)
All tests should use @WithMockOAuth2User to simulate an authenticated user
```
